src/analyser.cpp

```diff
@@ -1115,12 +1248,835 @@ void Analyser::AnalyserImpl::analyseEquationAst(const AnalyserEquationAstPtr &as
+            unitsString += (i == units.size() - 1) ? " and " : ", ";
+        }
+
+        unitsString += "'" + units[i] + "'";
+    }
+
+    return unitsString;
+}
+
+std::string Analyser::AnalyserImpl::expressionUnits(const AnalyserEquationAstPtr &ast,
+                                                    const UnitsMaps &unitsMaps,
+                                                    const UnitsMaps &userUnitsMaps,
+                                                    const UnitsMultipliers &unitsMultipliers)
+{
+    // Return a string version of the given AST and (user) units maps and units
+    // multipliers.
+
+    auto res = expression(ast, false) + " is ";
+    auto unitsString = expressionUnits(unitsMaps, unitsMultipliers);
+    auto userUnitsString = expressionUnits(userUnitsMaps);
+
+    if (userUnitsString.empty()) {
+        res += "'dimensionless'";
+    } else {
+        res += "in " + userUnitsString;
+
+        if (!unitsString.empty() && (unitsString != userUnitsString)) {
+            res += " (i.e. " + unitsString + ")";
+        }
+    }
+
+    return res;
+}
+
+void Analyser::AnalyserImpl::defaultUnitsMapsAndMultipliers(UnitsMaps &unitsMaps,
+                                                            UnitsMaps &userUnitsMaps,
+                                                            UnitsMultipliers &unitsMultipliers)
+{
+    // Default units maps and multipliers.
+
+    unitsMaps = {UnitsMap()};
+    userUnitsMaps = {UnitsMap()};
+    unitsMultipliers = {0.0};
+}
+
+void Analyser::AnalyserImpl::analyseEquationUnits(const AnalyserEquationAstPtr &ast,
+                                                  UnitsMaps &unitsMaps,
+                                                  UnitsMaps &userUnitsMaps,
+                                                  UnitsMultipliers &unitsMultipliers,
+                                                  Strings &issueDescriptions)
+{
+    // Analyse the units used with different MathML elements (table 2.1 of the
+    // CellML 2.0 normative specification; see https://bit.ly/3vBbyO5):
+    //  - Simple operands ('ci' and 'cn'; note: 'sep' is not relevant here): the
+    //    operand can have any unit.
+    //  - Basic structural (note: 'apply' is not relevant here):
+    //     - 'piecewise': the returned value of the different 'piece' and
+    //       'otherwise' statements should have the same unit .
+    //     - 'piece': the returned value can have any unit while the condition
+    //       should be dimensionless.
+    //     - 'otherwise': the returned value can have any unit.
+    //  - Relational operators ('eq', 'neq', 'gt', 'lt', 'geq' and 'leq'): the
+    //    two operands should have the same unit.
+    //  - Logical operators:
+    //     - 'and', 'or', 'xor': the two operands should be dimensionless.
+    //     - 'not': the operand should be dimensionless.
+    //  - Arithmetic operators:
+    //     - 'plus', 'times', and 'divide': the two operands should have
```

Comment:
times and divide don't need to have the same units...

Reply:
Oops! Fixed in commit 87a1847d.